rust/src/tx_builder.rs

```diff
@@ -175,110 +181,244 @@ impl TransactionBuilder {
         }
 
         Ok(())
     }
 
+    fn cip2_largest_first_by<F>(
+        &mut self,
+        available_inputs: &Vec<TransactionUnspentOutput>,
+        available_indices: &mut Vec<usize>,
+        input_total: &mut Value,
+        output_total: &mut Value,
+        by: F) -> Result<(), JsError>
+    where
+        F: Fn(&Value) -> Option<BigNum> {
+        let mut relevant_indices = available_indices.clone();
+        relevant_indices.retain(|i| by(&available_inputs[*i].output.amount).is_some());
+        // ordered in ascending order by predicate {by}
+        relevant_indices.sort_by_key(|i| by(&available_inputs[*i].output.amount).expect("filtered above"));
+
+        // iterate in decreasing order for predicate {by}
+        for i in relevant_indices.iter().rev() {
+            if by(input_total).unwrap_or(BigNum::zero()) >= by(output_total).expect("do not call on asset types that aren't in the output") {
+                break;
+            }
+            let input = &available_inputs[*i];
+            // differing from CIP2, we include the needed fees in the targets instead of just output values
+            let input_fee = self.fee_for_input(&input.output.address, &input.input, &input.output.amount)?;
+            self.add_input(&input.output.address, &input.input, &input.output.amount);
+            *input_total = input_total.checked_add(&input.output.amount)?;
+            *output_total = output_total.checked_add(&Value::new(&input_fee))?;
+            available_indices.swap_remove(available_indices.iter().position(|j| i == j).unwrap());
+        }
+
+        if by(input_total).unwrap_or(BigNum::zero()) < by(output_total).expect("do not call on asset types that aren't in the output") {
+            return Err(JsError::from_str("UTxO Balance Insufficient"));
+        }
+
+        Ok(())
+    }
+
+    fn cip2_random_improve_by<F>(
+        &mut self,
+        available_inputs: &Vec<TransactionUnspentOutput>,
+        available_indices: &mut BTreeSet<usize>,
+        input_total: &mut Value,
+        output_total: &mut Value,
+        by: F,
+        rng: &mut rand::rngs::ThreadRng) -> Result<(), JsError>
+    where
+        F: Fn(&Value) -> Option<BigNum> {
+        use rand::Rng;
+        // Phase 1: Random Selection
+        let mut relevant_indices = available_indices.iter()
+            .filter(|i| by(&available_inputs[**i].output.amount).is_some())
+            .cloned()
+            .collect::<Vec<usize>>();
+        let mut associated_indices: BTreeMap<TransactionOutput, Vec<usize>> = BTreeMap::new();
+        let mut outputs = self.outputs.0.iter()
+            .filter(|output| by(&output.amount).is_some())
+            .cloned()
+            .collect::<Vec<TransactionOutput>>();
+        outputs.sort_by_key(|output| by(&output.amount).expect("filtered above"));
+        for output in outputs.iter().rev() {
+            // TODO: how should we adapt this to inputs being associated when running for other assets?
```

Comment:
I'm not sure how significant this is for most wallets wrt the multiassets. We could do a halfway point and only consider this for the ADA-only stage (e.g. so we count the inputs added to satisfy the multiassets towards the ADA portion). Although it might be not much more work to just count inputs added in previous rounds (for other assets) against the outputs for subsequent rounds in general not just for ADA. It's a bit weird since random-improve is specified to do things per-output rather than for the total output required.